ivy.el

```diff
@@ -2649,12 +2649,19 @@ The first non-matching part is propertized:
         (i (1- (length str))))
     (catch 'done
       (while (>= i 0)
-        (when (equal (get-text-property i 'face str)
-                     '(completions-first-difference))
-          (throw 'done i))
+        (pcase (get-text-property i 'face str)
+          (`completions-first-difference (throw 'done i))
+          ((and (pred listp) (pred (member 'completions-first-difference)))
+           (throw 'done i)))
         (cl-decf i))
       (throw 'done (length str)))))
 
+(defun ivy--completion-prefix-offset (str md)
+  "Return the offset of the completion prefix in STR with its metadata MD."
+  (pcase (cdr (assoc 'category md))
+    (`file (length (file-name-directory str)))
+    (_ 0)))
```

Comment:
I imagine this doesn't handle completing e.g. `~/a/b/c` to `~/aaa/bbb/ccc`, right?

I think we can avoid the need for this function by instead taking advantage of the `base-size` already returned by `completion-all-completions`; stay tuned.